### PR TITLE
✨ Support bash -c "CMD" for docker RUN downloads-then-exec

### PR DESCRIPTION
**Pull request by laurentsimon**
* **Please check if the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
no support. see https://github.com/ossf/scorecard/pull/584#discussion_r656312099

* **What is the new behavior (if this is a feature change)?**
Support for `bash -c "CMD"` in docker `RUN`. see https://github.com/ossf/scorecard/pull/584#discussion_r656312099

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Other information**:

### CODE REVIEW

File: checks/shell_download_validate.go [hunk 1]
```diff
@@ -605,24 +618,64 @@ func recordFetchFileFromString(cmd string) (fmap map[string]bool, ok bool, err e
...
+	r, err = validateCommandIsNotFetchToStdinExecute(cmd, pathfn, logf)
+	if err != nil {
+		return false, err
+	} else if !r {
+		ret = false
+	}
 
-	return cmdValidated, nil
+	// TODO(laurent): add check for cat file | bash
+	// TODO(laurent): detect downloads of zip/tar files containing scripts.
```
Comment (oliverchang) [inline checks/shell_download_validate.go:666]: I think it's actually fine to stop here and not worry about these cases. Checking these accurately will be complex, making maintenance of this difficult and introducing more potential false positives.

File: checks/shell_download_validate.go [hunk 2]
```diff
@@ -493,91 +507,91 @@ func validateCommandIsNotNotSuFetchToStdinExecute(cmd, pathfn string,
...
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, ErrParsingShellCommand
+	}
+
+	cmdValidated := true
+	syntax.Walk(f, func(node syntax.Node) bool {
+		// Check if we're calling a file we previously downloaded.
+		if isExecuteFiles(node, cmd, pathfn, files, logf) {
+			cmdValidated = false
```
Comment (oliverchang) [inline checks/shell_download_validate.go:522]: do we need to continue walking in this case?

File: checks/shell_download_validate.go [hunk 3]
```diff
@@ -493,91 +507,91 @@ func validateCommandIsNotNotSuFetchToStdinExecute(cmd, pathfn string,
...
-		if err != nil {
+		if ok {
 			return false
 		}
-		var b bool
-		b, err = isSuFetchStdinExecute(node, cmd, pathfn, logf)
-		if err != nil {
+		s, kk := extractInterpreterCommandFromNode(node)
+		// nolinter
+		if kk {
```
Comment (oliverchang) [inline checks/shell_download_validate.go:570]: These variables (kk, cs, s) make this pretty confusing to understand. Can you rename them to more descriptive ones?

**Review by oliverchang — APPROVED** [review]
LGTM in addition to comments on your previous PR.

File: checks/shell_download_validate.go (hunk 3 above)
Comment (laurentsimon) [reply checks/shell_download_validate.go:570]: sure. I wanted to just do `cs, ok = extractInterpreterCommandFromNode(node)` but it does not work, ie `cs` seems to be a local alias, so I had to create these new variables.

Comment (laurentsimon) [reply checks/shell_download_validate.go]: renamed to `command, commandFound := extractInterpreterCommandFromNode(node)`

**Pull request merged**